src/snabbdom.ts

```diff
@@ -1,55 +1,90 @@
 
-type VNodeQueue = Array<VNode>;
+const { isArray } = Array;
 
-const emptyNode = vnode('', {}, [], undefined, undefined);
+const ELEMENT_NODE = 1, TEXT_NODE = 3, COMMENT_NODE = 8, DOCUMENT_FRAGMENT_NODE = 11;
 
-function sameVnode(vnode1: VNode, vnode2: VNode): boolean {
-  return vnode1.key === vnode2.key && vnode1.sel === vnode2.sel;
-}
+function isUndef(s: any): s is undefined { return s === undefined; }
+function isDef(s: any): boolean { return s !== undefined; }
 
-function isVnode(vnode: any): vnode is VNode {
-  return vnode.sel !== undefined;
-}
+type VNodeQueue = VNode[];
+
+const emptyNode: VNode = {
+  nt: 0,
+  sel: '',
+  data: {},
+  children: undefined,
+  text: undefined,
+  elm: undefined,
+  key: undefined,
+};
 
 type KeyToIndexMap = {[key: string]: number};
 
 type ArraysOf<T> = {
   [K in keyof T]: (T[K])[];
 }
 
-type ModuleHooks = ArraysOf<Module>;
+export interface PatchFunction {
+  (oldVnode: VNode, vnode: VNode): VNode;
+  children(parentElm: Node, oldCh: VNodes, newCh: VNodes): VNodes;
+}
+
+function defaultCompareFn(vnode1: VNode, vnode2: VNode): boolean {
+  return vnode1.nt === vnode2.nt && vnode1.key === vnode2.key && vnode1.sel === vnode2.sel;
+}
+
+function isVNode(vnode: any): vnode is VNode {
+  return vnode != null && vnode !== undefined;
+}
+
+function isElementVNode(vnode: VNode): vnode is VElement {
+  return vnode.nt === ELEMENT_NODE;
+}
 
-function createKeyToOldIdx(children: Array<VNode>, beginIdx: number, endIdx: number): KeyToIndexMap {
+function isFragmentVNode(vnode: VNode): vnode is VFragment {
+  return vnode.nt === DOCUMENT_FRAGMENT_NODE;
+}
+
+function isTextVNode(vnode: VNode): vnode is VText {
+  return vnode.nt === TEXT_NODE;
+}
+
+function isCommentVNode(vnode: VNode): vnode is VComment {
+  return vnode.nt === COMMENT_NODE;
+}
+
+function createKeyToOldIdx(children: VNodes, beginIdx: number, endIdx: number): KeyToIndexMap {
   let i: number, map: KeyToIndexMap = {}, key: Key | undefined, ch;
   for (i = beginIdx; i <= endIdx; ++i) {
     ch = children[i];
-    if (ch != null) {
+    if (isVNode(ch)) {
       key = ch.key;
       if (key !== undefined) map[key] = i;
     }
   }
   return map;
 }
 
-const hooks: (keyof Module)[] = ['create', 'update', 'remove', 'destroy', 'pre', 'post'];
-
-export {h} from './h';
-export {thunk} from './thunk';
+interface LifeCycleHooks {
+  pre: PreHook;
+  create: CreateHook;
+  update: UpdateHook;
+  destroy: DestroyHook;
+  remove: RemoveHook;
+  post: PostHook;
+}
 
-export function init(modules: Array<Partial<Module>>, domApi?: DOMAPI) {
-  let i: number, j: number, cbs = ({} as ModuleHooks);
+type ModuleHooks = ArraysOf<LifeCycleHooks>;
 
-  const api: DOMAPI = domApi !== undefined ? domApi : htmlDomApi;
+const hooks: (keyof LifeCycleHooks)[] = ['create', 'update', 'remove', 'destroy', 'pre', 'post'];
 
+export function init(modules: Array<Partial<Module>>, api: DOMAPI,
```

Comment:
the signature of the `init` has changed a little bit, it now takes 2 required arguments and one optional:

* modules (same as before)
* api (same as before, but it is now required)
* compareFn (optional argument that fallbacks to `defaultCompareFn` from above)

the reasoning behind the required second argument is that if you're bundling the snabbdom core library, it should work as a generic diffing library with no affiliations to the DOM at all, and the interaction with the real tree should be proxied by the api provided as a second argument as an indirection.